File: packages/sdk-python/jc_kami/_ed25519.py
"""Ed25519 验签。

SDK 承诺零依赖，但标准库没有 Ed25519，所以这里按 RFC 8032 自带一份验签实现。
装了 cryptography 就优先用它（更快、经过更多审计），没有就退回纯 Python 实现——
验签在整个更新流程里只跑一次，纯 Python 的几十毫秒完全不影响体验。

只实现 verify：SDK 永远不需要签名，私钥只存在于服务端。
"""

import hashlib

__all__ = ["verify"]

# ---- 曲线参数（RFC 8032 §5.1） ----
_P = 2**255 - 19
_L = 2**252 + 27742317777372353535851937790883648493
_D = -121665 * pow(121666, _P - 2, _P) % _P
_SQRT_M1 = pow(2, (_P - 1) // 4, _P)
_By = 4 * pow(5, _P - 2, _P) % _P


def _sha512(b):
    return hashlib.sha512(b).digest()


def _recover_x(y, sign):
    """由 y 与符号位还原 x，失败返回 None（表示这不是曲线上的点）。"""
    if y >= _P:
        return None
    xx = (y * y - 1) * pow(_D * y * y + 1, _P - 2, _P) % _P
    if xx == 0:
        return None if sign else 0
    x = pow(xx, (_P + 3) // 8, _P)
    if (x * x - xx) % _P != 0:
        x = x * _SQRT_M1 % _P
    if (x * x - xx) % _P != 0:
        return None
    if x % 2 != sign:
        x = _P - x
    return x


_Bx = _recover_x(_By, 0)
# 扩展坐标 (X, Y, Z, T)，避免每次加法都求逆元
_B = (_Bx, _By, 1, _Bx * _By % _P)
_IDENTITY = (0, 1, 1, 0)


def _point_add(p, q):
    x1, y1, z1, t1 = p
    x2, y2, z2, t2 = q
    a = (y1 - x1) * (y2 - x2) % _P
    b = (y1 + x1) * (y2 + x2) % _P
    c = 2 * t1 * t2 * _D % _P
    d = 2 * z1 * z2 % _P
    e, f, g, h = b - a, d - c, d + c, b + a
    return (e * f % _P, g * h % _P, f * g % _P, e * h % _P)


def _scalar_mult(p, e):
    q = _IDENTITY
    while e > 0:
        if e & 1:
            q = _point_add(q, p)
        p = _point_add(p, p)
        e >>= 1
    return q


def _point_equal(p, q):
    x1, y1, z1, _ = p
    x2, y2, z2, _ = q
    # 投影坐标下相等要交叉相乘比较，不能直接比 X/Y
    return (x1 * z2 - x2 * z1) % _P == 0 and (y1 * z2 - y2 * z1) % _P == 0


def _decompress(data):
    """解压 32 字节点表示，失败返回 None。"""
    if len(data) != 32:
        return None
    y = int.from_bytes(data, "little")
    sign = y >> 255
    y &= (1 << 255) - 1
    x = _recover_x(y, sign)
    if x is None:
        return None
    return (x, y, 1, x * y % _P)


# 曲线上 8 个小阶点的压缩表示（与 libsodium 的黑名单一致）。
# 用这些点当公钥时，能构造出「对任意消息都验得过」的签名。
# 正常流程不会碰到它们，但客户端一旦把公钥读成全零之类的坏值，
# 攻击者就能借此让恶意包通过验签——所以宁可直接拒绝。
_SMALL_ORDER_POINTS = frozenset(
    bytes.fromhex(h)
    for h in (
        "0100000000000000000000000000000000000000000000000000000000000000",
        "ecffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
        "0000000000000000000000000000000000000000000000000000000000000000",
        "0000000000000000000000000000000000000000000000000000000000000080",
        "26e8958fc2b227b045c3f489f2ef98f0d5dfac05d3c63339b13802886d53fc05",
        "c7176a703d4dd84fba3c0b760d10670f2a2053fa2c39ccc64ec7fd7792ac037a",
        "26e8958fc2b227b045c3f489f2ef98f0d5dfac05d3c63339b13802886d53fc85",
        "c7176a703d4dd84fba3c0b760d10670f2a2053fa2c39ccc64ec7fd7792ac03fa",
    )
)


def _verify_pure(public_key, message, signature):
    if len(public_key) != 32 or len(signature) != 64:
        return False
    if bytes(public_key) in _SMALL_ORDER_POINTS:
        return False
    a = _decompress(public_key)
    if a is None:
        return False
    rs = signature[:32]
    r = _decompress(rs)
    if r is None:
        return False
    s = int.from_bytes(signature[32:], "little")
    # 拒绝非规范的 S，避免签名可延展
    if s >= _L:
        return False
    k = int.from_bytes(_sha512(rs + public_key + message), "little") % _L
    # 校验 [S]B == R + [k]A
    return _point_equal(_scalar_mult(_B, s), _point_add(r, _scalar_mult(a, k)))


def verify(public_key, message, signature):
    """验证 Ed25519 签名。public_key 为裸 32 字节，signature 为 64 字节。"""
    # 前置检查放在分岔之前，保证走 cryptography 和走纯 Python 的结论一致
    if len(public_key) != 32 or len(signature) != 64:
        return False
    if bytes(public_key) in _SMALL_ORDER_POINTS:
        return False
    try:
        from cryptography.exceptions import InvalidSignature
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

        try:
            Ed25519PublicKey.from_public_bytes(public_key).verify(signature, message)
            return True
        except (InvalidSignature, ValueError):
            return False
    except ImportError:
        return _verify_pure(public_key, message, signature)

File: packages/sdk-csharp/src/JcKami.Sdk/Ed25519.cs
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Security.Cryptography;

namespace JcKami
{
    /// <summary>
    /// Ed25519 验签（RFC 8032）。
    ///
    /// .NET 直到 8.0 都没有内置 Ed25519（10.0 才补上 System.Security.Cryptography.Ed25519），
    /// 而本 SDK 承诺零依赖、且要覆盖 netstandard2.0（.NET Framework 4.6.1+），
    /// 所以这里自带一份实现。验签在整个更新流程里只跑一次，
    /// BigInteger 版的几毫秒完全不影响体验。
    ///
    /// 只实现 verify：客户端永远不需要签名，私钥只存在于服务端。
    /// </summary>
    internal static class Ed25519
    {
        // ---- 曲线参数（RFC 8032 §5.1） ----
        private static readonly BigInteger P =
            BigInteger.Pow(2, 255) - 19;

        private static readonly BigInteger L =
            BigInteger.Pow(2, 252) + BigInteger.Parse("27742317777372353535851937790883648493");

        private static readonly BigInteger D = Mod(-121665 * Inv(121666));
        private static readonly BigInteger SqrtM1 = BigInteger.ModPow(2, (P - 1) / 4, P);

        private static readonly Point B = MakeBasePoint();
        private static readonly Point Identity = new Point(0, 1, 1, 0);

        /// <summary>
        /// 曲线上 8 个小阶点的压缩表示（与 libsodium 的黑名单一致）。
        ///
        /// 用这些点当公钥时，能构造出「对任意消息都验得过」的签名。
        /// 正常流程不会碰到它们，但客户端一旦把公钥读成全零之类的坏值，
        /// 攻击者就能借此让恶意包通过验签 —— 所以宁可直接拒绝。
        /// </summary>
        private static readonly HashSet<string> SmallOrderPoints = new HashSet<string>(StringComparer.Ordinal)
        {
            "0100000000000000000000000000000000000000000000000000000000000000",
            "ecffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
            "0000000000000000000000000000000000000000000000000000000000000000",
            "0000000000000000000000000000000000000000000000000000000000000080",
            "26e8958fc2b227b045c3f489f2ef98f0d5dfac05d3c63339b13802886d53fc05",
            "c7176a703d4dd84fba3c0b760d10670f2a2053fa2c39ccc64ec7fd7792ac037a",
            "26e8958fc2b227b045c3f489f2ef98f0d5dfac05d3c63339b13802886d53fc85",
            "c7176a703d4dd84fba3c0b760d10670f2a2053fa2c39ccc64ec7fd7792ac03fa",
        };

        /// <summary>验证 Ed25519 签名。publicKey 为裸 32 字节，signature 为 64 字节。</summary>
        public static bool Verify(byte[] publicKey, byte[] message, byte[] signature)
        {
            if (publicKey == null || signature == null || message == null) return false;
            if (publicKey.Length != 32 || signature.Length != 64) return false;
            if (SmallOrderPoints.Contains(DeviceIdentity.ToHex(publicKey))) return false;

            Point a;
            if (!Decompress(publicKey, out a)) return false;

            var rs = new byte[32];
            Buffer.BlockCopy(signature, 0, rs, 0, 32);
            Point r;
            if (!Decompress(rs, out r)) return false;

            var sBytes = new byte[32];
            Buffer.BlockCopy(signature, 32, sBytes, 0, 32);
            var s = FromLittleEndian(sBytes);
            // 拒绝非规范的 S，避免签名可延展
            if (s >= L) return false;

            byte[] hash;
            using (var sha = SHA512.Create())
            {
                var buffer = new byte[32 + 32 + message.Length];
                Buffer.BlockCopy(rs, 0, buffer, 0, 32);
                Buffer.BlockCopy(publicKey, 0, buffer, 32, 32);
                Buffer.BlockCopy(message, 0, buffer, 64, message.Length);
                hash = sha.ComputeHash(buffer);
            }

            var k = Mod(FromLittleEndian(hash), L);

            // 校验 [S]B == R + [k]A
            return PointEqual(ScalarMult(B, s), PointAdd(r, ScalarMult(a, k)));
        }

        // ================================================================ 群运算

        private struct Point
        {
            // 扩展坐标 (X, Y, Z, T)，避免每次加法都求逆元
            public readonly BigInteger X, Y, Z, T;

            public Point(BigInteger x, BigInteger y, BigInteger z, BigInteger t)
            {
                X = x; Y = y; Z = z; T = t;
            }
        }

        private static Point MakeBasePoint()
        {
            var by = Mod(4 * Inv(5));
            BigInteger bx;
            if (!RecoverX(by, 0, out bx))
                throw new InvalidOperationException("Ed25519 基点初始化失败");
            return new Point(bx, by, BigInteger.One, Mod(bx * by));
        }

        private static Point PointAdd(Point p, Point q)
        {
            var a = Mod((p.Y - p.X) * (q.Y - q.X));
            var b = Mod((p.Y + p.X) * (q.Y + q.X));
            var c = Mod(2 * p.T * q.T * D);
            var d = Mod(2 * p.Z * q.Z);
            BigInteger e = b - a, f = d - c, g = d + c, h = b + a;
            return new Point(Mod(e * f), Mod(g * h), Mod(f * g), Mod(e * h));
        }

        private static Point ScalarMult(Point p, BigInteger e)
        {
            var q = Identity;
            while (e > 0)
            {
                if (!e.IsEven) q = PointAdd(q, p);
                p = PointAdd(p, p);
                e >>= 1;
            }
            return q;
        }

        private static bool PointEqual(Point p, Point q)
        {
            // 投影坐标下相等要交叉相乘比较，不能直接比 X/Y
            return Mod(p.X * q.Z - q.X * p.Z).IsZero
                && Mod(p.Y * q.Z - q.Y * p.Z).IsZero;
        }

        /// <summary>由 y 与符号位还原 x。返回 false 表示这不是曲线上的点。</summary>
        private static bool RecoverX(BigInteger y, int sign, out BigInteger x)
        {
            x = BigInteger.Zero;
            if (y >= P) return false;

            var xx = Mod((y * y - 1) * BigInteger.ModPow(Mod(D * y * y + 1), P - 2, P));
            if (xx.IsZero) return sign == 0;

            x = BigInteger.ModPow(xx, (P + 3) / 8, P);
            if (!Mod(x * x - xx).IsZero) x = Mod(x * SqrtM1);
            if (!Mod(x * x - xx).IsZero) { x = BigInteger.Zero; return false; }

            if ((x.IsEven ? 0 : 1) != sign) x = P - x;
            return true;
        }

        /// <summary>解压 32 字节点表示。</summary>
        private static bool Decompress(byte[] data, out Point point)
        {
            point = Identity;
            if (data == null || data.Length != 32) return false;

            var raw = new byte[32];
            Buffer.BlockCopy(data, 0, raw, 0, 32);
            var sign = (raw[31] >> 7) & 1;
            raw[31] &= 0x7F;

            var y = FromLittleEndian(raw);
            BigInteger x;
            if (!RecoverX(y, sign, out x)) return false;

            point = new Point(x, y, BigInteger.One, Mod(x * y));
            return true;
        }

        // ================================================================ 工具

        private static BigInteger Inv(BigInteger value) => BigInteger.ModPow(Mod(value), P - 2, P);

        private static BigInteger Mod(BigInteger value) => Mod(value, P);

        private static BigInteger Mod(BigInteger value, BigInteger modulus)
        {
            var r = value % modulus;
            return r.Sign < 0 ? r + modulus : r;
        }

        /// <summary>小端字节序转无符号大整数。</summary>
        private static BigInteger FromLittleEndian(byte[] bytes)
        {
            // BigInteger(byte[]) 按有符号解释，末尾补一个 0 字节才能保证得到正数
            var padded = new byte[bytes.Length + 1];
            Buffer.BlockCopy(bytes, 0, padded, 0, bytes.Length);
            return new BigInteger(padded);
        }
    }
}
